integration-cli/docker_api_swarm_test.go

```diff
@@ -445,6 +445,43 @@ func (s *DockerSwarmSuite) TestApiSwarmServicesStateReporting(c *check.C) {
 		}
 	}
 }
+func (s *DockerSwarmSuite) TestApiSwarmLeaderElection(c *check.C) {
+	// Create 3 nodes
+	d1 := s.AddDaemon(c, true, true)
+	d2 := s.AddDaemon(c, true, true)
+	d3 := s.AddDaemon(c, true, true)
+
+	// pick out the leader
+	var leader *SwarmDaemon
+	for _, d := range []*SwarmDaemon{d1, d2, d3} {
+		if d1.getNode(c, d.NodeID).ManagerStatus.Leader {
+			leader = d
+			break
+		}
+	}
+
+	// stop the leader
+	leader.Stop()
+
+	// wait for an election to occur
+	time.Sleep(time.Second * 2)
+
+	// add the old leader back
+	leader.Start()
+
+	var newleader *SwarmDaemon
+	// check to see that the leader has changed to one of the other nodes
+	for _, d := range []*SwarmDaemon{d1, d2, d3} {
+		if d2.getNode(c, d.NodeID).ManagerStatus.Leader {
+			newleader = d
+			break
+		}
+	}
+
+	// assert that we have a new leader, and that it is not the old leader
```

Comment:
I have a revision of this that I'm working on where I just iterate through the two remaining nodes until one of this is the leader.

It is guaranteed in this test that we have a new leader that isn't the old leader though, because the old leader is stopped while the election should be occurring.